cEP-0025.md

```diff
@@ -0,0 +1,275 @@
+
+## Demonstrating use of PyFlakesASTBear by creating a fixes bear
+
+A hassle free solution to the problem of removing all the future imports
+present in the source code.
+
+Here is a prototype for the NoFutureImportBear that uses PyFlakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+import astor
+
+class RemoveFutureImports(ast.NodeTransformer):
+    def visit_ImportFrom(self, node):
+        try:
+            if node.module == '__future__':
+                return None
+        except AttributeError:
+            pass
+        return node
+
+class NoFutureImportBear(LocalBear):
+    """
+    Uses pyflakes-enhance-AST to remove future imports
+    """
+    BEAR_DEPS = {PyFlakesASTBear}
+
+    def remove_future_imports(self, file, lineno):
+        """
+        Removes all ImportFrom nodes from the input line
+        """
+        diff = Diff(file)
+        line = file[lineno - 1]
+        tree = ast.parse(line)
+        RemoveFutureImports().visit(tree)
+        replacement = '; '.join(astor.to_source(tree).splitlines())
+        if replacement == '':
+            diff.delete_line(lineno)
+        else:
+            diff.modify_line(lineno, replacement)
+
+        return diff
+
+    def run(self, filename, file,
+            dependency_results=dict()
+            ):
+        """
+        Uses PyFlakesASTBear to get only FutureImportation
+        """
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            affected_lines = set()
+            for node in result.get_nodes(result.module_scope[0],
+                                         FutureImportation):
+                lineno = result.get_node_location(node)[0]
+                affected_lines.add(lineno)
+
+        for line in affected_lines:
+            corrected = self.remove_future_imports(file, line)
+            yield Result.from_values(
+                origin=self,
+                message=('Future import/imports found'),
+                file=filename,
+                diffs={filename: corrected},
+                line=line)
+```
+
+## A generic implementation for detecting `__future__` imports
+
+This implementation doesn't requires coala to be installed and is
+designed so that it can be used in other tools like flake8. This
+implementation will be housed in `pyflakes-generic-plugins`.
+
+Here is a prototype for the NoFutureImport that uses pyflakes-AST:
+
+```python
+# Imports not mentioned to maintain brevity
+__version__ = '0.1'
+
+CODE = 'F482'
+
+
+class NoFutureImport(object):
+    name = 'no_future'
+    version = __version__
+
+    def __init__(self, tree, filename):
+        self.tree = tree
+        self.filename = filename
+        self.checker = None
+
+    def set_pyflakes_data(self, checker):
+        self.checker = checker
+
+    def run(self):
+        if self.checker is None:
+            self.set_pyflakes_data(Checker(self.tree, self.filename))
+        module_scope = list(filter(lambda scope: isinstance(scope,
```

Comment:
you can also move this into a property getter to move the `run` look a lot simpler